crates/chain/src/canonical_iter.rs

```diff
@@ -0,0 +1,319 @@
+use crate::collections::{hash_map, HashMap, HashSet, VecDeque};
+use crate::tx_graph::{CanonicalTx, TxAncestors, TxDescendants, TxNode};
+use crate::{Anchor, ChainOracle, ChainPosition, TxGraph};
+use alloc::boxed::Box;
+use alloc::collections::BTreeSet;
+use alloc::sync::Arc;
+use bdk_core::BlockId;
+use bitcoin::{Transaction, Txid};
+
+/// A set of canonical transactions.
+pub type CanonicalSet<A> = HashMap<Txid, (Arc<Transaction>, CanonicalReason<A>)>;
+
+/// Iterates over canonical txs.
+pub struct CanonicalIter<'g, A, C> {
+    tx_graph: &'g TxGraph<A>,
+    chain: &'g C,
+    chain_tip: BlockId,
+
+    pending_anchored: Box<dyn Iterator<Item = (Txid, Arc<Transaction>, &'g BTreeSet<A>)> + 'g>,
+    pending_last_seen: Box<dyn Iterator<Item = (Txid, Arc<Transaction>, u64)> + 'g>,
+    pending_remaining: VecDeque<(Txid, Arc<Transaction>, u32)>,
+
+    canonical: CanonicalSet<A>,
+    not_canonical: HashSet<Txid>,
+
+    queue: VecDeque<Txid>,
+}
+
+impl<'g, A: Anchor, C: ChainOracle> CanonicalIter<'g, A, C> {
+    /// Constructs [`CanonicalIter`].
+    pub fn new(tx_graph: &'g TxGraph<A>, chain: &'g C, chain_tip: BlockId) -> Self {
+        let anchors = tx_graph.all_anchors();
+        let pending_anchored = Box::new(
+            tx_graph
+                .txids_by_descending_anchor_height()
+                .filter_map(|(_, txid)| Some((txid, tx_graph.get_tx(txid)?, anchors.get(&txid)?))),
+        );
+        let pending_last_seen = Box::new(
+            tx_graph
+                .txids_by_descending_last_seen()
+                .filter_map(|(last_seen, txid)| Some((txid, tx_graph.get_tx(txid)?, last_seen))),
+        );
+        Self {
+            tx_graph,
+            chain,
+            chain_tip,
+            pending_anchored,
+            pending_last_seen,
+            pending_remaining: VecDeque::new(),
+            canonical: HashMap::new(),
+            not_canonical: HashSet::new(),
+            queue: VecDeque::new(),
+        }
+    }
+
+    /// Whether this transaction is already visited.
+    fn is_visited(&self, txid: Txid) -> bool {
+        self.canonical.contains_key(&txid) || self.not_canonical.contains(&txid)
```

Comment:
`is_visited`  is a bit of a lie. `scan_anchors` is visiting the tx but it doesn't always make it `is_visited`. This should be `is_canonicalized`.

Reply:
Changed in the latest push.